File: client-support/src/main/java/io/github/ruedigerk/contractfirst/generator/client/ApiClientIoException.java
package io.github.ruedigerk.contractfirst.generator.client;

import java.io.IOException;
import java.util.Optional;

/**
 * Thrown when an IOException occurs during a request of the API client.
 */
public class ApiClientIoException extends ApiClientException {

  private final ApiRequest request;
  private final IncompleteResponse incompleteResponse;

  public ApiClientIoException(String message, ApiRequest request, IOException cause) {
    super(toMessage(message, request), cause);
    this.request = request;
    incompleteResponse = null;
  }

  public ApiClientIoException(String message, IncompleteResponse incompleteResponse, IOException cause) {
    super(toMessage(message, incompleteResponse), cause);
    request = incompleteResponse.getRequest();
    this.incompleteResponse = incompleteResponse;
  }

  private static String toMessage(String message, ApiRequest request) {
    return message + ", for: " + request.getMethod() + " " + request.getUrl();
  }

  private static String toMessage(String message, IncompleteResponse incompleteResponse) {
    ApiRequest request = incompleteResponse.getRequest();
    return message + 
        ", for " + request.getMethod() + " " + request.getUrl() + 
        ", status=" + incompleteResponse.getStatusCode() + " " + incompleteResponse.getHttpStatusMessage();
  }

  /**
   * Returns a description of the request that lead to the IOException.
   */
  public ApiRequest getRequest() {
    return request;
  }

  /**
   * Return the incomplete response from the server, if available.
   */
  public Optional<IncompleteResponse> getIncompleteResponse() {
    return Optional.ofNullable(incompleteResponse);
  }
}

File: client-support/src/main/kotlin/io/github/ruedigerk/contractfirst/generator/client/ApiClientIoException.kt
package io.github.ruedigerk.contractfirst.generator.client

import java.io.IOException

/**
 * Thrown when an IOException occurs during a request of the API client.
 */
class ApiClientIoException : ApiClientException {

  /**
   * The description of the request that lead to the IOException.
   */
  val request: ApiRequest?

  /**
   * The incomplete response from the server, if available.
   */
  @Suppress("MemberVisibilityCanBePrivate")
  val incompleteResponse: IncompleteResponse?

  constructor(message: String, request: ApiRequest, cause: IOException?) : super(toMessage(message, request), cause) {
    this.request = request
    incompleteResponse = null
  }

  constructor(message: String, incompleteResponse: IncompleteResponse, cause: IOException?) : super(toMessage(message, incompleteResponse), cause) {
    request = incompleteResponse.request
    this.incompleteResponse = incompleteResponse
  }

  private companion object {

    private fun toMessage(message: String, request: ApiRequest): String {
      return message + ", for: " + request.method + " " + request.url
    }

    private fun toMessage(message: String, incompleteResponse: IncompleteResponse): String {
      val request = incompleteResponse.request
      return message +
        ", for " + request.method + " " + request.url +
        ", status=" + incompleteResponse.statusCode + " " + incompleteResponse.httpStatusMessage
    }
  }
}
